File: lang/ruby/src/fun/vac/data_structures/ArrayList.rb
module ArrayList

    class ArrayList

        @@DEFAULT_CAPACITY = 64

        public
        def initialize(physicalSize = 0)
            @data = nil
            @logicalSize = 0
            @physicalSize = @@DEFAULT_CAPACITY

            if physicalSize > 1
                @physicalSize = physicalSize
            end
            @data = Array.new(@physicalSize)
        end

        public
        def size()
            @logicalSize
        end

        def isEmpty()
            @logicalSize.zero?
        end

        def get(index)
            if index.negative? || index >= @logicalSize
                raise "[PANIC - IndexOutOfBounds]"
            end

            @data[index]
        end

        def set(index, element)
            if index.negative? || index >= @logicalSize
                raise "[PANIC - IndexOutOfBounds]"
            end

            @data[index] = element
        end

        def insert(index, element)
            if index.negative? || index > @logicalSize
                raise "[PANIC - IndexOutOfBounds]"
            end

            if @logicalSize == @physicalSize
                newCapacity = @@DEFAULT_CAPACITY

                if @physicalSize >= @@DEFAULT_CAPACITY
                    newCapacity = @physicalSize + (@physicalSize >> 1)
                end

                temp = Array.new(newCapacity)

                0.upto(@logicalSize - 1) do |i|
                    temp[i] = @data[i]
                end

                @data = temp
                @physicalSize = newCapacity
            end

            @logicalSize.downto(index + 1) do |i|
                @data[i] = @data[i - 1]
            end

            @data[index] = element

            @logicalSize += 1
        end

        def remove(index)
            if index.negative? || index >= @logicalSize
                raise "[PANIC - IndexOutOfBounds]"
            end

            (index + 1).upto(@logicalSize - 1) do |i|
                @data[i - 1] = @data[i]
            end

            @logicalSize -= 1

            @data[@logicalSize] = nil
        end

        def front()
            get(0)
        end

        def back()
            get(@logicalSize - 1)
        end

        def prepend(element)
            insert(0, element)
        end

        def append(element)
            insert(@logicalSize, element)
        end

        def poll()
            remove(0)
        end

        def eject()
            remove(@logicalSize - 1)
        end

        def capacity()
            @physicalSize
        end

        def shrink()
            temp = Array.new(@logicalSize)

            0.upto(@logicalSize - 1) do |i|
                temp[i] = @data[i]
            end

            @data = temp
            @physicalSize = @logicalSize
        end
    end
end

File: lang/python/src/fun/vac/data_structures/ArrayList.py
class ArrayList:
    DEFAULT_CAPACITY = 64

    def __init__(self, physicalSize: int = 0):
        self.data = None
        self.logicalSize = 0
        self.physicalSize = self.DEFAULT_CAPACITY

        if physicalSize > 1:
            self.physicalSize = physicalSize
        self.data = [0] * self.physicalSize

    def size(self) -> int:
        return self.logicalSize

    def isEmpty(self) -> bool:
        return self.logicalSize == 0

    def get(self, index: int) -> int:
        if index < 0 or index >= self.logicalSize:
            raise RuntimeError("[PANIC - IndexOutOfBounds]")

        return self.data[index]

    def set(self, index: int, element: object) -> None:
        if index < 0 or index >= self.logicalSize:
            raise RuntimeError("[PANIC - IndexOutOfBounds]")

        self.data[index] = element

    def insert(self, index: int, element: object) -> None:
        if index < 0 or index > self.logicalSize:
            raise RuntimeError("[PANIC - IndexOutOfBounds]")

        if self.logicalSize == self.physicalSize:
            newCapacity = self.DEFAULT_CAPACITY

            if self.physicalSize >= self.DEFAULT_CAPACITY:
                newCapacity = self.physicalSize + (self.physicalSize >> 1)

            temp = [0] * newCapacity

            for i in range(self.logicalSize):
                temp[i] = self.data[i]

            self.data = temp
            self.physicalSize = newCapacity

        for i in range(self.logicalSize, index, -1):
            self.data[i] = self.data[i - 1]

        self.data[index] = element

        self.logicalSize += 1

    def remove(self, index: int) -> None:
        if index < 0 or index >= self.logicalSize:
            raise RuntimeError("[PANIC - IndexOutOfBounds]")

        for i in range(index + 1, self.logicalSize):
            self.data[i - 1] = self.data[i]

        self.logicalSize -= 1

        self.data[self.logicalSize] = None

    def front(self) -> int:
        return self.get(0)

    def back(self) -> int:
        return self.get(self.logicalSize - 1)

    def prepend(self, element: object) -> None:
        self.insert(0, element)

    def append(self, element: object) -> None:
        self.insert(self.logicalSize, element)

    def poll(self) -> None:
        self.remove(0)

    def eject(self) -> None:
        self.remove(self.logicalSize - 1)

    def capacity(self) -> int:
        return self.physicalSize

    def shrink(self) -> None:
        temp = [0] * self.logicalSize

        for i in range(self.logicalSize):
            temp[i] = self.data[i]

        self.data = temp
        self.physicalSize = self.logicalSize
